### ISSUE TITLE
referencing resources with for_each causes invalid index errors (terraform 0.12.24)

### ISSUE THREAD

**Issue by axelgobletbdr** [hashicorp/terraform#24594]
### Terraform Version

```
0.12.24
```

### Terraform Configuration Files

```hcl
locals {
  service_name = "wifi_data_pipeline"
  lambda_folder = "${path.module}/src/wifi_data_pipeline/api"
  lambda_handler = "lambda_handler"
  default_lambdas = toset(["schema",
    "scanner_metadata",
    "weak_signal_filter",
    "stationary_device_filter",
    "scanner_group",
    "aggregation",
    "imputation",
    "group_conversion_factor",
    "format_timedeltas",
    "dwell_time"
  ])
}

data "archive_file" "default_lambda" {
  for_each = local.default_lambdas
  output_path = "${path.module}/build/${each.value}.zip"
  type = "zip"
  source_file = "${local.lambda_folder}/${each.value}.py"
}

resource "aws_s3_bucket_object" "default_lambda" {
  for_each = local.default_lambdas
  bucket = aws_s3_bucket.functions.bucket
  key = "${each.value}.zip"
  source = data.archive_file.default_lambda[each.key].output_path
  etag = data.archive_file.default_lambda[each.key].output_md5
}

resource "aws_lambda_function" "default_lambda" {
  for_each = local.default_lambdas
  function_name = "${local.service_name}_${each.value}"
  handler = "${each.value}.${local.lambda_handler}"
  role = aws_iam_role.processed.arn
  runtime = element(tolist(aws_lambda_layer_version.main.compatible_runtimes), 0)
  memory_size = 3008
  s3_bucket = aws_s3_bucket.functions.bucket
  s3_key = aws_s3_bucket_object.default_lambda[each.value].key
  source_code_hash = data.archive_file.default_lambda[each.value].output_base64sha256
  layers = [aws_lambda_layer_version.main.arn]
  timeout = 900
}
```

### Expected Behavior

Terraform creates each of the three resources for each item in `default_lambdas`

### Actual Behavior

Terraform raises an "Error: Invalid Index":

```
Error: Invalid index
  on modules/wifi_data_pipeline/main.tf line 313, in resource "aws_lambda_function" "default_lambda":
 313:   s3_key = aws_s3_bucket_object.default_lambda[each.value].key
    |----------------
    | aws_s3_bucket_object.default_lambda is object with 1 attribute "dwell_time"
    | each.value is "imputation"
```

This error happens for every item in `default_lambdas`, except for `dwell_time`

### Steps to Reproduce

`terraform apply`

### Additional Context

These resources reside in a sub-module of my terraform root module

The same error appears when I change `default_lambdas` to a map like this:

```
{
    schema = 1
    scanner_metadata = 1
    weak_signal_filter = 1
    stationary_device_filter = 1
    scanner_group = 1
    aggregation = 1
    imputation = 1
    group_conversion_factor = 1
    format_timedeltas = 1
    dwell_time = 1
}
```

I am currently using `count` as a workaround, which does not cause any problems

### References

These issues seem to cover the problem in earlier Terraform versions. 

- #23077
- #22407

**Comment by danieldreier** [comment]
@axelgobletbdr we've recently adopted a process of building simple reproduction cases for issues, and it has helped us fix bugs much faster. I tried to put together a simplified reproduction case using `null_resource` in https://github.com/danieldreier/terraform-issue-reproductions/tree/master/24594 but am not seeing this behavior even though it looks like the essential elements you're describing are there. I'm using null resources because it makes local testing way easier when we don't have to provision real infrastructure.

If you can help make a PR against my test case above where you can reproduce this, it would help me move this issue forward. As-is, the reproduction case you described isn't immediately usable because I don't have the rest of your terraform config. Once we have a more isolated reproduction case, I can escalate this to engineers.

**Comment by axelgobletbdr** [comment]
Hi Daniel, 

I see that you did not use tf 1.12.24, maybe the problem is already fixed in your later versions?

**Comment by axelgobletbdr** [comment]
Furthermore, in my example, there are multiple data and resource objects that depend on each other and have a `for_each`. Maybe that is the source of the error?

**Comment by nijave** [comment]
For anyone that runs into this... Problem appears fixed in 0.14 but `lookup` might be a workaround if you're still on an older version
